core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -62,29 +61,24 @@
 
 (def ^:dynamic *column->pushdown-bloom* {})
 
-(defn- filter-pushdown-bloom-page-idxs [^IMetadataManager metadata-manager buf-key ^String col-name ^RoaringBitmap page-idxs]
+(defn filter-pushdown-bloom-page-idx-pred ^IntPredicate [^IMetadataManager metadata-manager ^String col-name
+                                                         {:keys [trie-file ^RelationReader trie-rdr] :as _trie-match} ^IntPredicate page-idx-pred]
   (if-let [^MutableRoaringBitmap pushdown-bloom (get *column->pushdown-bloom* (symbol col-name))]
     ;; would prefer this `^long` to be on the param but can only have 4 params in a primitive hinted function in Clojure
-    @(meta/with-metadata metadata-manager buf-key
-       (util/->jfn
-         (fn [^ITableMetadata table-metadata]
-           (let [metadata-rdr (.metadataReader table-metadata)
-                 bloom-rdr (-> (.structKeyReader metadata-rdr "columns")
-                               (.listElementReader)
-                               (.structKeyReader "bloom"))
-                 filtered-page-idxs (RoaringBitmap.)]
-             (.forEach page-idxs
-                       (reify org.roaringbitmap.IntConsumer
-                         (accept [_ page-idx]
-                           (when-let [bloom-vec-idx (.rowIndex table-metadata col-name page-idx)]
-                             (when (and (not (nil? (.getObject bloom-rdr bloom-vec-idx)))
-                                        (MutableRoaringBitmap/intersects pushdown-bloom
-                                                                         (bloom/bloom->bitmap bloom-rdr bloom-vec-idx)))
-                               (.add filtered-page-idxs page-idx))))))
-
-             (when-not (.isEmpty filtered-page-idxs)
-               filtered-page-idxs)))))
-    page-idxs))
+    (let [^ITableMetadata table-metadata (.tableMetadata metadata-manager trie-rdr trie-file)
+          metadata-rdr (.metadataReader table-metadata)
+          bloom-rdr (-> (.structKeyReader metadata-rdr "columns")
+                        (.listElementReader)
+                        (.structKeyReader "bloom"))]
+      (reify IntPredicate
+        (test [_ page-idx]
+          (boolean
+           (when (.test page-idx-pred page-idx)
```

Comment:
I don't know if this "predicate stacking" has any performance implications.

Reply:
oh, I see, it's not just two predicates, it's used within a `reduce`. same probably still applies unless we're talking hundreds?

Reply:
if you're concerned, rather than a reduce here, I'd create an IntPredicate for each (i.e. map) and then write a single 'and' intpredicate (if it doesn't already exist) which tests all of them in order (short-circuiting if possible)

Reply:
Good point. `IntPredicate` has a `(.and ...)` method.